src/client/mod.rs

```diff
@@ -255,20 +132,134 @@ impl TeeConnection {
+                    .find(|ext| ext.extn_id == RaTlsQuoteExtension::OID)
+                    .ok_or_else(|| Error::General("Failed get quote in certificate".into()))?
+                    .extn_value
+                    .as_bytes();
+
+                trace!("Get collateral bytes!");
+
+                let collateral = exts
+                    .iter()
+                    .find(|ext| ext.extn_id == RaTlsCollateralExtension::OID)
+                    .and_then(|ext| {
+                        serde_json::from_slice::<Collateral>(ext.extn_value.as_bytes())
+                            .map_err(|e| {
+                                debug!("Failed to get collateral in certificate {e:?}");
+                                trace!(
+                                    "Failed to get collateral in certificate {:?}",
+                                    String::from_utf8_lossy(ext.extn_value.as_bytes())
+                                );
+                            })
+                            .ok()
+                    });
+
+                if collateral.is_none() {
+                    debug!("Failed to get collateral in certificate");
+                }
+
+                let quote = Quote::try_from_bytes(quote_bytes).map_err(|e| {
+                    Error::General(format!("Failed get quote in certificate {e:?}"))
+                })?;
+
+                if &quote.report_body.reportdata[..32] != hash.as_slice() {
+                    error!("Report data mismatch");
+                    return Err(Error::General("Report data mismatch".to_string()));
+                } else {
                     info!(
-                        "Checked or set server certificate public key hash `{}`",
-                        hex::encode(&hash[..32])
+                        "Report data matches `{}`",
+                        hex::encode(&quote.report_body.reportdata[..32])
                     );
-                    Ok(rustls::client::danger::ServerCertVerified::assertion())
-                } else {
-                    error!("Server certificate does not match expected certificate");
-                    Err(rustls::Error::General(
-                        "Server certificate does not match expected certificate".to_string(),
-                    ))
                 }
+
+                let current_time: i64 = time::SystemTime::now()
+                    .duration_since(time::UNIX_EPOCH)
+                    .unwrap()
+                    .as_secs() as _;
+
+                let QuoteVerificationResult {
+                    collateral_expired,
+                    result,
+                    quote,
+                    advisories,
+                    earliest_expiration_date,
+                    ..
+                } = verify_quote_with_collateral(quote_bytes, collateral.as_ref(), current_time)
+                    .unwrap();
+
+                if collateral_expired || !matches!(result, sgx_ql_qv_result_t::SGX_QL_QV_RESULT_OK)
```

Comment:
Maybe Rust n00b, but why not `result !=  sgx_ql_qv_result_t::SGX_QL_QV_RESULT_OK`?

Reply:
Thanks! Must be some leftover.

Reply:
fixed